tools/RepositoryManagement/DevHome.RepositoryManagement/Services/EnhanceRepositoryService.cs

```diff
@@ -0,0 +1,170 @@
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using DevHome.Common.Services;
+using DevHome.Common.TelemetryEvents.RepositoryManagement;
+using DevHome.Common.Windows.FileDialog;
+using DevHome.Customization.ViewModels;
+using DevHome.Telemetry;
+using FileExplorerSourceControlIntegration;
+using Microsoft.UI.Xaml;
+using Microsoft.Windows.DevHome.SDK;
+using Serilog;
+using Windows.Foundation.Collections;
+using Windows.Storage;
+
+namespace DevHome.RepositoryManagement.Services;
+
+public class EnhanceRepositoryService
+{
+    private const string ErrorEventName = "DevHome_EnhanceRepositoryError_Event";
+
+    private readonly ILogger _log = Log.ForContext("SourceContext", nameof(EnhanceRepositoryService));
+
+    private readonly FileExplorerViewModel _sourceControlRegistrar;
+
+    private readonly Window _window;
+
+    private readonly ITelemetry _telemetry;
+
+    private readonly List<IExtensionWrapper> _repositorySourceControlProviders;
+
+    public EnhanceRepositoryService(
+        FileExplorerViewModel sourceControlRegistrar,
+        Window window,
+        IExtensionService extensionService)
+    {
+        _sourceControlRegistrar = sourceControlRegistrar;
+        _telemetry = TelemetryFactory.Get<ITelemetry>();
+        _window = window;
+        _repositorySourceControlProviders = extensionService.GetInstalledExtensionsAsync(ProviderType.LocalRepository).Result.ToList();
+    }
+
+    /// <summary>
+    /// Gets the location of an existing repository and hook up the path to both
+    /// the source control provider and file explorer.
+    /// </summary>
+    /// <returns>Path to the new repository.  String.Empty for any errors</returns>
+    public async Task<(string RepositoryLocation, Guid sourceControlClassId)> SelectRepositoryAndMakeItEnhanced()
+    {
+        var maybeRepositoryPath = await GetRepositoryLocationFromUser();
+        var sourceControlId = await MakeRepositoryEnhanced(maybeRepositoryPath);
+        return (maybeRepositoryPath, sourceControlId);
+    }
+
+    public async Task<Guid> MakeRepositoryEnhanced(string repositoryLocation)
+    {
+        _sourceControlRegistrar.AddRepositoryAlreadyOnMachine(repositoryLocation);
+        var sourceControlId = await AssignSourceControlToPath(repositoryLocation);
+        return sourceControlId;
+    }
+
+    public string GetLocalBranchName(string repositoryLocation)
+    {
+        try
+        {
+            var repository = new LibGit2Sharp.Repository(repositoryLocation);
+            return repository.Head.FriendlyName;
+        }
+        catch (Exception ex)
+        {
+            _log.Error(ex, $"Error when getting branch name");
+        }
+
+        return string.Empty;
+    }
+
+    public string GetRepositoryUrl(string repositoryLocation)
+    {
+        try
+        {
+            var repository = new LibGit2Sharp.Repository(repositoryLocation);
```

Comment:
Ditto for this leaking `Repository`.

Is there a reason you're using Git-specific stuff instead of adding it to the extension SDK for general use?

Reply:
> adding it to the extension SDK for general use?

Not enough time and a low concentration of non-git repositories.  The process to
1. Design DevHome to work with arbitrary columns
2. API review to allow extensions to provide
    1. Column names
    2. Column data
3. Implement the changes for DevHome and the two existing extensions.

More than I had time for.  If needed, APIs can be added in the future to support a generic Repository Management screen.